README.rst
================
 CERNServiceXML
================

.. image:: https://travis-ci.org/inveniosoftware/cernservicexml.svg?branch=master
    :target: https://travis-ci.org/inveniosoftware/cernservicexml
.. image:: https://coveralls.io/repos/inveniosoftware/cernservicexml/badge.svg?branch=master
    :target: https://coveralls.io/r/inveniosoftware/cernservicexml
.. image:: https://pypip.in/v/cernservicexml/badge.svg
    :target: https://pypi.python.org/pypi/cernservicexml/
.. image:: https://pypip.in/d/cernservicexml/badge.svg
    :target: https://pypi.python.org/pypi/cernservicexml/

About
=====

CERNServiceXML is a small library to generate a CERN XSLS Service XML
and sending it to a remove service.

Installation
============

CERNServiceXML is on PyPI so all you need is: ::

    pip install cernservicexml

Documentation
=============

Documentation is readable at http://cernservicexml.readthedocs.org or can be
build using Sphinx: ::

    pip install Sphinx
    python setup.py build_sphinx

Testing
=======

Running the test suite is as simple as: ::

    python setup.py test

or, to also show code coverage: ::

    ./run-tests.sh


CHANGES
Changes
=======

Version 0.2.1 (released 2015-07-24)

- Relaxes Requests library version requirement from v2.3 to v1.1 to support
  CERN SLC6/CentOS 7 distributions. Note, Python 3.4 requires Requests v2+.
- Adds Travis test matrix to test multiple versions of the Requests library.
- Standardises software name (CERNServiceXML) and package name (cernservicexml)
  following usual inveniosoftware practices.
- Fixes reST headline formatting in `RELEASE-NOTES.rst`.

Version 0.2.0 (released 2015-07-16)

- Adds support for service status parameter in ``ServiceDocument``.
- Deprecates use of service availability percentage in ``ServiceDocument``.
  Please use service status parameter instead.
- Changes documentation theme to the standard non-Flask RTD one.
- Initial release of Docker configuration suitable for local developments.
  `docker-compose build` rebuilds the image, `docker-compose run --rm web
  python setup.py test` runs the test suite.

Version 0.1.2 (released 2015-06-17)

- Adds support for long (Python 2) integers.

Version 0.1.1 (released 2015-06-17)

- Fixes issues with POST request for sending the service XML to XSLS.
- Adds doctests to ensure documentation examples works.
- Adds XML validation tests according to XSD Schema.

Version 0.1.0 (released 2015-06-16)

- Initial public release.


run-tests.sh
#!/bin/sh
#
# This file is part of CERNServiceXML
# Copyright (C) 2015 CERN.
#
# CERNServiceXML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.

pep257 cernservicexml && \
sphinx-build -qnNW docs docs/_build/html && \
python setup.py test && \
sphinx-build -qnNW -b doctest docs docs/_build/doctest


tests/helpers.py
# -*- coding: utf-8 -*-
#
# This file is part of CERN Service XML
# Copyright (C) 2015 CERN.
#
# CERN Service XML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.


"""Test helpers."""

from __future__ import absolute_import, print_function, unicode_literals

import functools
import sys

PY34 = sys.version_info[0:2] >= (3, 4)


def import_httpretty():
    """Import HTTPretty and monkey patch Python 3.4 issue.

    See https://github.com/gabrielfalcao/HTTPretty/pull/193 and
    as well as https://github.com/gabrielfalcao/HTTPretty/issues/221.
    """
    if not PY34:
        import httpretty
    else:
        import socket
        old_SocketType = socket.SocketType

        import httpretty
        from httpretty import core

        def sockettype_patch(f):
            @functools.wraps(f)
            def inner(*args, **kwargs):
                f(*args, **kwargs)
                socket.SocketType = old_SocketType
                socket.__dict__['SocketType'] = old_SocketType
            return inner

        core.httpretty.disable = sockettype_patch(
            httpretty.httpretty.disable
        )

    return httpretty


tests/test_document.py
# -*- coding: utf-8 -*-
#
# This file is part of CERNServiceXML
# Copyright (C) 2015 CERN.
#
# CERNServiceXML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.

"""Service document tests."""

from __future__ import absolute_import, print_function, unicode_literals

from datetime import datetime
from decimal import Decimal
from os.path import dirname, join

import pytest
from cernservicexml import ServiceDocument, Status
from cernservicexml._compat import StringIO, long_type
from lxml import etree

schema = etree.XMLSchema(file=join(dirname(__file__), 'xsls_schema.xsd'))


def validate_xsd(xmlstr):
    """Assert if xml is valid according to XSD schema."""
    schema.assertValid(etree.parse(StringIO(xmlstr)))
    return True


def test_creation_simple():
    """Test creation of a service document."""
    doc = ServiceDocument('myserviceid')
    assert doc.service_id == 'myserviceid'
    assert isinstance(doc.timestamp, datetime)
    assert doc.status == Status.available

    assert validate_xsd(doc.to_xml())
    assert doc.to_xml() == \
        '<serviceupdate xmlns="http://sls.cern.ch/SLS/XML/update">' \
        '<id>myserviceid</id>' \
        '<status>available</status>' \
        '<timestamp>{0}</timestamp>' \
        '</serviceupdate>'.format(doc.timestamp.isoformat())

    dt = datetime(2015, 1, 1, 0, 0, 0)
    doc = ServiceDocument('anotherid', timestamp=dt,
                          status=Status.degraded)

    assert doc.to_xml() == \
        '<serviceupdate xmlns="http://sls.cern.ch/SLS/XML/update">' \
        '<id>anotherid</id>' \
        '<status>degraded</status>' \
        '<timestamp>2015-01-01T00:00:00</timestamp>' \
        '</serviceupdate>'
    assert validate_xsd(doc.to_xml())


def test_availablity_to_status():
    """Test creation of a service document."""
    dt = datetime(2015, 1, 1, 0, 0, 0)
    doc = ServiceDocument('anotherid', timestamp=dt, availability=99)
    assert doc.status == Status.available
    assert doc.availability == 99
    assert validate_xsd(doc.to_xml())

    doc = ServiceDocument('anotherid', timestamp=dt, availability=60)
    assert doc.status == Status.degraded
    assert doc.availability == 60
    assert validate_xsd(doc.to_xml())

    doc = ServiceDocument('anotherid', timestamp=dt, availability=20)
    assert doc.status == Status.unavailable
    assert doc.availability == 20
    assert validate_xsd(doc.to_xml())

    with pytest.raises(AssertionError):
        doc = ServiceDocument('anid', timestamp=dt, status=Status.unavailable,
                              availability=20)

    doc = ServiceDocument('anotherid', timestamp=dt, status='available')
    assert doc.status == Status.available
    assert doc.availability == 100
    assert validate_xsd(doc.to_xml())

    doc = ServiceDocument('anotherid', timestamp=dt, status='degraded')
    assert doc.status == Status.degraded
    assert doc.availability == 50
    assert validate_xsd(doc.to_xml())

    doc = ServiceDocument('anotherid', timestamp=dt, status='unavailable')
    assert doc.status == Status.unavailable
    assert doc.availability == 0
    assert validate_xsd(doc.to_xml())


def test_creation_attrs():
    """Test non-mandatory attributes."""
    dt = datetime(2015, 1, 1, 0, 0, 0)
    doc = ServiceDocument(
        'myid',
        timestamp=dt,
        availabilitydesc='My description',
        contact='info@example.org',
        webpage='http://example.org',
        availabilityinfo='Extra info',
    )

    assert doc.to_xml() == \
        '<serviceupdate xmlns="http://sls.cern.ch/SLS/XML/update">' \
        '<id>myid</id>' \
        '<status>available</status>' \
        '<timestamp>2015-01-01T00:00:00</timestamp>' \
        '<availabilitydesc>My description</availabilitydesc>' \
        '<contact>info@example.org</contact>' \
        '<webpage>http://example.org</webpage>' \
        '<availabilityinfo>Extra info</availabilityinfo>' \
        '</serviceupdate>'
    assert validate_xsd(doc.to_xml())


def test_outofbounds():
    """Test out of bounds values."""
    with pytest.raises(AssertionError):
        ServiceDocument(None)
    with pytest.raises(AssertionError):
        ServiceDocument(1234)
    with pytest.raises(AssertionError):
        ServiceDocument('id', timestamp='1234')
    with pytest.raises(AssertionError):
        ServiceDocument('id', status=100)
    with pytest.raises(AssertionError):
        ServiceDocument('id', status='not-a-status')
    doc = ServiceDocument('id')
    with pytest.raises(AssertionError):
        doc.add_numericvalue(1234, 'val')
    with pytest.raises(AssertionError):
        doc.add_numericvalue('name', 1234, desc=1234)
    with pytest.raises(AssertionError):
        doc.add_numericvalue('name', 'astring', desc=1234)
    # Remove in v0.3 (availability removed)
    with pytest.raises(AssertionError):
        ServiceDocument('id', availability=-1)
    with pytest.raises(AssertionError):
        ServiceDocument('id', availability=101)
    with pytest.raises(AssertionError):
        ServiceDocument('id', availability='100')


def test_numericvalue():
    """Test adding numeric data to service."""
    dt = datetime(2015, 1, 1, 0, 0, 0)
    doc = ServiceDocument('myid', timestamp=dt)
    doc.add_numericvalue('val1', 1234)
    doc.add_numericvalue('val2', 12.34)
    doc.add_numericvalue('val3', Decimal("0.1"))
    doc.add_numericvalue('val4', Decimal("0.1") + Decimal("0.2"))
    doc.add_numericvalue('val5', 1234, desc="a desc")
    doc.add_numericvalue('val6', long_type(1234))

    assert doc.to_xml() == \
        '<serviceupdate xmlns="http://sls.cern.ch/SLS/XML/update">' \
        '<id>myid</id>' \
        '<status>available</status>' \
        '<timestamp>2015-01-01T00:00:00</timestamp>' \
        '<data>' \
        '<numericvalue name="val1">1234</numericvalue>' \
        '<numericvalue name="val2">12.34</numericvalue>' \
        '<numericvalue name="val3">0.1</numericvalue>' \
        '<numericvalue name="val4">0.3</numericvalue>' \
        '<numericvalue desc="a desc" name="val5">1234</numericvalue>' \
        '<numericvalue name="val6">1234</numericvalue>' \
        '</data>' \
        '</serviceupdate>'
    assert validate_xsd(doc.to_xml())


cernservicexml/_compat.py
# -*- coding: utf-8 -*-
#
# This file is part of CERNServiceXML
# Copyright (C) 2015 CERN.
#
# CERNServiceXML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.

"""Python 2/3 compatibility module."""

from __future__ import absolute_import, print_function, unicode_literals

import functools
import sys

# Useful for very coarse version differentiation.
PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3
PY34 = sys.version_info[0:2] >= (3, 4)

if PY3:  # pragma: no cover
    string_types = str,
    text_type = str
    binary_type = bytes
    from io import StringIO
    long_type = int
else:    # pragma: no cover
    string_types = basestring,
    text_type = unicode
    binary_type = str
    from StringIO import StringIO
    long_type = long


def import_httpretty():  # pragma: no cover
    """Import HTTPretty and monkey patch Python 3.4 issue.

    See https://github.com/gabrielfalcao/HTTPretty/pull/193 and
    as well as https://github.com/gabrielfalcao/HTTPretty/issues/221.
    """
    if not PY34:
        import httpretty
    else:
        import socket
        old_SocketType = socket.SocketType

        import httpretty
        from httpretty import core

        def sockettype_patch(f):
            @functools.wraps(f)
            def inner(*args, **kwargs):
                f(*args, **kwargs)
                socket.SocketType = old_SocketType
                socket.__dict__['SocketType'] = old_SocketType
            return inner

        core.httpretty.disable = sockettype_patch(
            httpretty.httpretty.disable
        )

    return httpretty

__all__ = (
    'binary_type',
    'import_httpretty',
    'string_types',
    'StringIO',
    'text_type',
)


cernservicexml/publisher.py
# -*- coding: utf-8 -*-
#
# This file is part of CERNServiceXML
# Copyright (C) 2015 CERN.
#
# CERNServiceXML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.

"""Publishing of service documents.

.. testsetup:: *

   from cernservicexml._compat import import_httpretty
   httpretty = import_httpretty()
   httpretty.enable()
   httpretty.register_uri(httpretty.POST, 'http://xsls.cern.ch',
                          body="", status=200)

.. testcleanup::

   httpretty.disable()


Usage:

>>> from cernservicexml import ServiceDocument, XSLSPublisher
>>> doc = ServiceDocument('myserviceid')
>>> XSLSPublisher.send(doc)
<Response [200]>
"""

from __future__ import absolute_import, print_function, unicode_literals

import requests

from ._compat import StringIO
from .document import ServiceDocument


class XSLSPublisher(object):

    """Publish a service document to an XSLS service."""

    @classmethod
    def send(cls, document, api_url='http://xsls.cern.ch'):
        """Send service document to XSLS."""
        assert isinstance(document, ServiceDocument)
        return requests.post(api_url, files=dict(
            file=StringIO(document.to_xml())))


cernservicexml/document.py
# -*- coding: utf-8 -*-
#
# This file is part of CERNServiceXML
# Copyright (C) 2015 CERN.
#
# CERNServiceXML is free software; you can redistribute it and/or modify
# it under the terms of the Revised BSD License; see LICENSE file for
# more details.

"""Creation of XSLS service documents.

Usage:

>>> from cernservicexml import ServiceDocument, Status
>>> doc = ServiceDocument('zenodo', status=Status.available)
>>> doc.add_numericvalue('users', 1000, desc="Number of users")
>>> xml = doc.to_xml()
"""

from __future__ import absolute_import, print_function, unicode_literals

import warnings
import xml.etree.ElementTree as ET
from datetime import datetime
from decimal import Decimal

from ._compat import binary_type, long_type, string_types, text_type


class Status(object):

    """Representation of availability status."""

    available = 'available'
    """Service is available for all users."""

    degraded = 'degraded'
    """Part of the service is not be working or some users are affected."""

    unavailable = 'unavailable'
    """Service is unavailable for all users."""


class ServiceDocument(object):

    """XSLS Service Document class.

    :param service_id: A unique service id.
    :param availability: An integer 0-100 indicating availability of
        service (*deprecated*, use ``status`` instead). Default: 100.
    :param status: Status of the service. Allowed values: ``available``,
        ``degraded`` or ``unavailable``. Default: ``available``.
    :param timestamp: Timestamp when the availability was calculated. Must
        be a datetime object. Default: now.
    :param availabilitydesc: Information about the availability.
    :param contact: Service contacts a string.
    :param webpage: Service website (any URI)
    :param availabilityinfo: Information about the service. Can contain
        HTML.

    .. versionchanged:: 0.2
       Added ``status`` parameter. Deprecated ``availability``
       parameter.
    """

    STATUSES = [Status.available, Status.degraded, Status.unavailable]
    """Allowed status values."""

    def __init__(self, service_id, status=None, timestamp=None,
                 availabilitydesc=None, contact=None, webpage=None,
                 availabilityinfo=None,
                 availability=None):
        """Initialize a service document."""
        if timestamp is not None:
            assert isinstance(timestamp, datetime)
        assert isinstance(service_id, string_types)

        if status is not None:
            assert status in self.STATUSES
            if availability is not None:
                raise AssertionError(
                    "Status and availability cannot be specified at the same "
                    "time. Please use only status."
                )
        elif availability is not None:
            warnings.warn(
                "Keyword argument 'availability' is deprecated and will be "
                "removed in v0.3. Please use keyword argument 'status' "
                "instead.", DeprecationWarning)
            assert isinstance(availability, int)
            assert availability >= 0 and availability <= 100
            self._availablity = availability
            if availability >= 0 and availability < 40:
                status = Status.unavailable
            elif availability >= 40 and availability < 70:
                status = Status.degraded
            elif availability >= 70 and availability <= 100:
                status = Status.available
        else:
            status = Status.available

        self.service_id = service_id
        self.status = status
        self.timestamp = timestamp or datetime.now()
        self.availabilitydesc = availabilitydesc
        self.contact = contact
        self.webpage = webpage
        self.availabilityinfo = availabilityinfo
        self.data = []

    @property
    def availability(self):
        """Get status as legacy availability percentage.

        .. deprecated:: 0.2
           Use of availability as percentage is deprecated. Please use
           ``status`` instead. Attribute will be removed in v0.3.
        """
        warnings.warn(
            "Use of availability attribute is deprecated. Please use status "
            "attribute instead.", DeprecationWarning)

        # If availability was set via constructor return that value.
        if hasattr(self, '_availablity'):
            return self._availablity

        # Otherwise, compute availability from status.
        if self.status == Status.available:
            return 100
        elif self.status == Status.degraded:
            return 50
        else:
            return 0

    def add_numericvalue(self, name, value, desc=None):
        """Add a numeric value metric.

        :param name: Numeric value id.
        :param value: An integer, float or Decimal object.
        :param desc: A description.

        """
        assert isinstance(name, string_types)
        assert isinstance(value, int) or isinstance(value, float) or \
            isinstance(value, Decimal) or isinstance(value, long_type)
        if desc is not None:
            assert isinstance(desc, string_types)
        self.data.append(dict(name=name, desc=desc, value=value))

    def to_xml(self):
        """Serialize service document to XML."""
        root = ET.Element('serviceupdate',
                          xmlns="http://sls.cern.ch/SLS/XML/update")
        ET.SubElement(root, 'id').text = self.service_id
        ET.SubElement(root, 'status').text = text_type(self.status)
        ET.SubElement(root, 'timestamp').text = self.timestamp.isoformat()

        for e in ['availabilitydesc', 'contact', 'webpage',
                  'availabilityinfo']:
            val = getattr(self, e)
            if val is not None:
                ET.SubElement(root, e).text = text_type(val)

        if self.data:
            data = ET.SubElement(root, 'data')

            for d in self.data:
                attrs = dict(name=d['name'])
                if d['desc']:
                    attrs['desc'] = d['desc']
                ET.SubElement(data, 'numericvalue', attrib=attrs).text = \
                    text_type(d['value'])

        res = ET.tostring(root)
        if isinstance(res, binary_type):
            res = res.decode('utf-8')

        return res
